servant-server/src/Servant/Server/Internal/RoutingApplication.hs

```diff
@@ -155,32 +159,111 @@ data Delayed :: * -> * where
+                -- auth data was succesfully extracted from the request
+                Right aData -> do
+                    eUsr <- checkAuth authProtection aData
+                    case eUsr of
+                        -- we failed to authenticate the user
+                        Left uError -> unauthHandler (onUnauthenticated authProtection) (subserver authProtection) uError aData
+                        -- user was authenticated
+                        Right usr ->
+                            (return . Route . subserver authProtection) (returnHandler usr)
+    in Delayed captures method newAuth body (\_ y _ -> Route y)
+
+-- | Delayed auth checker for Strict Missing and Strict Unauthentication
+addAuthCheckSS :: Delayed (AuthProtected IO ServantErr 'Strict mError 'Strict uError auth usr (usr -> a))
+               -> IO (RouteResult (Either mError auth))
+               -> Delayed a
+addAuthCheckSS = genAuthCheck (\(StrictMissing handler) _ e -> FailFatal <$> handler e)
+                              (\(StrictUnauthenticated handler) _ e a -> FailFatal <$> handler e a)
+                              id
+
+-- | Delayed auth checker for Strict Missing and Lax Unauthentication
+addAuthCheckSL :: Delayed (AuthProtected IO ServantErr 'Strict mError 'Lax uError auth usr (Either uError usr -> a))
+               -> IO (RouteResult (Either mError auth))
+               -> Delayed a
+addAuthCheckSL = genAuthCheck (\(StrictMissing handler) _ e -> FailFatal <$> handler e)
+                              (\(LaxUnauthenticated) cont e _ -> (return . Route . cont) (Left e))
+                              Right
+
+
+-- | Delayed auth checker for Lax Missing and Strict Unauthentication
+addAuthCheckLS :: Delayed (AuthProtected IO ServantErr 'Lax mError 'Strict uError auth usr (Either mError usr -> a))
+               -> IO (RouteResult (Either mError auth))
+               -> Delayed a
+addAuthCheckLS = genAuthCheck (\(LaxMissing) cont e -> (return . Route . cont) (Left e))
+                              (\(StrictUnauthenticated handler) _ e a -> FailFatal <$> handler e a)
+                              Right
+
+-- | Delayed auth checker for Lax Missing and Lax Unauthentication
+addAuthCheckLL :: Delayed (AuthProtected IO ServantErr 'Lax mError 'Lax uError auth usr (Either (Either mError uError) usr -> a))
+               -> IO (RouteResult (Either mError auth))
+               -> Delayed a
+addAuthCheckLL = genAuthCheck (\(LaxMissing) cont e -> (return . Route . cont) (Left (Left e)))
+                              (\(LaxUnauthenticated) cont e _ -> (return . Route . cont) (Left (Right e)))
+                              Right
+
+-- | Add an auth check by supplying OnMissing policies and OnUnauthenticated policies.
+addAuthCheck :: SAuthPolicy mPolicy
+             -> SAuthPolicy uPolicy
+             -> Delayed (AuthProtected IO ServantErr mPolicy mError uPolicy uError auth usr (AuthDelayedReturn mPolicy mError uPolicy uError usr -> a))
```

Comment:
this allows someone to use `addAuthCheck` like:

```
addAuthCheck SStrict SLax
```

instead of relying on the convention of ending each combinator with `S` or `L`.